### ISSUE TITLE
High memory use

### ISSUE THREAD

**Issue by CavalcanteLeo** [rs/SDWebImage#1544]
SDWebImageManager *manager = [SDWebImageManager sharedManager];
    [manager downloadImageWithURL:post.imageURL options:0 progress:^(NSInteger receivedSize, NSInteger expectedSize) 
    {
        // float progress = (float)receivedSize/expectedSize;
        // [cell.imageViewFoto updateProgress:progress];
    }
    completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL *imageURL) {
        if (image) {
            [cell.imageViewFoto setupImageView:post indexPath:indexPath isGrid:NO  image:image];
        }
    }];

<img width="1392" alt="captura de tela 2016-04-30 as 15 34 24" src="https://cloud.githubusercontent.com/assets/1735944/14937809/28317e34-0ee9-11e6-948d-d9532923b7d9.png">

**Comment by mythodeia** [comment]
whats the size of the image you are trying to download?

**Comment by CavalcanteLeo** [comment]
about 500kb each...
showing in a collectionview, about 12 images at same time...
The problem occurs when i scroll the collectionview, and request more and more and more images...

Maybe i have to disable the memory cache and only store in disk.. but i have no idea how to del with it



![2016-04-30 16 16 21](https://cloud.githubusercontent.com/assets/1735944/14938078/5519bb0e-0eef-11e6-930e-90434d99f979.png)

**Comment by mythodeia** [comment]
did you try setting these two:
```
[[SDImageCache sharedImageCache] setShouldDecompressImages:NO];
[[SDWebImageDownloader sharedDownloader] setShouldDecompressImages:NO];
```
by the way, whats the highest resolution of those images?

**Comment by CavalcanteLeo** [comment]
No, i havent, i will try this code... 


The API returns me a lot of images already resized, so i choose witch one to use, depending of the iPhone resolution(100x100, 200x200, 500x500, 1000x1000, 1500x1500, 2000x2000)
in this particular case, im using the 500x500 ( but the print of the collectionview with the images was using 100x100 for testing)

**Comment by CavalcanteLeo** [comment]
Btw, where should ai use this code? appdelegate or viewcontroller?

**Comment by mythodeia** [comment]
should be before you make your requests. try the appdelegate

**Comment by CavalcanteLeo** [comment]
On viewcontroller nothing changed... still increasing memory

**Comment by CavalcanteLeo** [comment]
even on appdelegate, nothing have changed too... =(

**Comment by CavalcanteLeo** [comment]
"thumbnail": {
          "height": 100,
          "url": "https://timehi.s3-sa-east-1.amazonaws.com/post/160c6ef0a2d2db2a6502.jpg",
          "width": 100
        }

the 100x100 pixels image is about 4kb...

and increase a lot of memory

**Comment by mythodeia** [comment]
why are you using SDWebImageManager's ```downloadImageWithURL:...``` and not ```[cell.imageView sd_setImageWithURL:...```
can you try the demo project? do you see the same increase in memory there too?

**Comment by CavalcanteLeo** [comment]
i cant show the image as soon as the dowload finish, i have to implement some logic before with the image...

**Comment by mythodeia** [comment]
if you perform extra actions on the image you could use [this](https://github.com/rs/SDWebImage/blob/469d8bbfd8b32ee5aba09d618fe2e6444fadb217/SDWebImage/SDWebImageManager.h#L127)

did you try the demo project? does it show the same memory increase?

**Comment by CavalcanteLeo** [comment]
i will, try right now

**Comment by CavalcanteLeo** [comment]
it still increasing memory with the code bellow:


    [cell.image sd_setImageWithURL:downloadImg completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, NSURL *imageURL) {
        if (image) {
            BOOL isGrid = YES;
            if (self.headerProfile.segmentedControl.selectedSegmentIndex == 1) {
                isGrid = NO;
                cell.image.delegate = self;
            }else{
                cell.image.delegate = nil;
            }
            [cell.image setupImageView:post indexPath:indexPath isGrid:isGrid image:image];
        }
    }];

**Comment by CavalcanteLeo** [comment]
also have checked all strong and weak properties referencies... just to double check if the ARC is releasing all object correctly, but it's all right with it

**Comment by mythodeia** [comment]
try the demo project with your API urls but without doing anything extra to the image after download. do you still see any increase in memory?

**Comment by CavalcanteLeo** [comment]
I have to go now, but i will try in about 2 hours, tks by now

**Comment by bpoplauschi** [comment]
@CavalcanteLeo any updates? Any chance you are using animated (GIF) images?

**Comment by CavalcanteLeo** [comment]
by changing setShouldDecompressImages to YES, in some benchmarks, my app performance went from 600mb to about 50mb

    [[SDWebImageDownloader sharedDownloader] setShouldDecompressImages:NO];
    [[SDImageCache sharedImageCache] setShouldDecompressImages:NO];

and then from 50mb to about 5mb with this:

    [[SDImageCache sharedImageCache] setShouldCacheImagesInMemory:NO];

![captura de tela 2016-05-02 as 15 58 58](https://cloud.githubusercontent.com/assets/1735944/14964515/cff95e2a-107e-11e6-8667-9daf7e6cc0ac.png)



also set some other properties (no impact changed):

    [[SDImageCache sharedImageCache] setMaxMemoryCost:1024 * 1024 * 1];
    [[SDImageCache sharedImageCache] setMaxCacheAge:3600 * 24 * 7];


and on each viecontroller e clean the cache

    -(void)viewDidDisappear:(BOOL)animated{
        [super viewDidDisappear:animated];
        [self ckearImageCache];
    }
    -(void)didReceiveMemoryWarning{
        [super didReceiveMemoryWarning];
        [self ckearImageCache];
    }
    -(void)ckearImageCache{
        SDImageCache *imageCache = [SDImageCache sharedImageCache];
        [imageCache clearMemory];
    }


but now, my problem is app performance, as i set ShouldDecompressImages to NO. 
When i scroll my collection view to fast, my app freezes a bit

**Comment by mime29** [comment]
Did you try to activate the address sanitizer in your scheme? I have myself a crash because of sdwebimage when i activate it.

**Comment by ynotski** [comment]
We have also been monitoring an issue with RAM usage being way to high. We upgraded to v3.7.3 and followed the example above. e.g.:

**Comment by SpaceStar2016** [comment]
I am using 3.7.3 and did the code like    
 [[SDImageCache sharedImageCache] setShouldDecompressImages:YES];
    [[SDWebImageDownloader sharedDownloader] setShouldDecompressImages:YES];
    [[SDImageCache sharedImageCache] setShouldCacheImagesInMemory:NO]
it didn't work for me. How to solve the problem??

**Comment by ynotski** [comment]
@SpaceStar2016 
`setShouldCacheImagesInMemory:` is a method not available in v 3.7.3, so I am fairly sure you're not running v 3.7.3.

If you're using `cocoapods` to add dependencies to your project, change your `podfile` to:

pod 'SDWebImage', '3.7.3'

then run `pod install`

Remove the line:
`[[SDImageCache sharedImageCache] setShouldCacheImagesInMemory:NO]` or you will get compile errors.

Hopefully this will help.

**Comment by SpaceStar2016** [comment]
@ynotski I am sure I am using 3.7.3 I can code [[SDImageCache sharedImageCache] setShouldCacheImagesInMemory:NO] in my project.
![0f0bbc40-6e71-4b40-884e-f201ed225846](https://cloud.githubusercontent.com/assets/17593419/16140023/c1aaff7a-347f-11e6-84bb-644cd9c7f513.png)

**Comment by bpoplauschi** [comment]
I think this issue is similar to #538 #586 and #1354.
#787 and 00bf467 should give you a good solution for dealing with large images. Together with the GIF handled by FLAnimatedImage, this ticket should be done now.

Closing. Let's see how this is with 4.0.0

**Comment by michal-gumny-nomtek** [comment]
This is closed but just want to add that I experience must higher memory usage in 4.0.0-beta2 that in mentioned here 3.7.3. In both cases with each image download memory is rising and not falling back later

**Comment by dipkasyap** [comment]
Having same issue here, I am on xcode 9.4, swift 4.X, and tested on ios 11(Simulator) 
![screen shot 2018-06-12 at 6 21 50 pm](https://user-images.githubusercontent.com/12591229/41290838-c022b36c-6e6d-11e8-93dd-11e0ad419338.png)

**Comment by smindia1988** [comment]
After lots of R&D and applying all kind of options for SDImageCache, I have finalized below options to set to avoid crash on heavy images load with the Best performance on RAM cache load:

Hope this will help you all !!!

Add below code into your AppDelegate class:

import SDWebImage

```
SDImageCache.shared().config.maxCacheAge = 3600 * 24 * 7 //1 Week
        SDImageCache.shared().maxMemoryCost = 1024 * 1024 * 20 //Aprox 20 images
        //SDImageCache.shared().config.shouldCacheImagesInMemory = false //Default True => Store images in RAM cache for Fast performance
        SDImageCache.shared().config.shouldDecompressImages = false
        SDWebImageDownloader.shared().shouldDecompressImages = false
        SDImageCache.shared().config.diskCacheReadingOptions = NSData.ReadingOptions.mappedIfSafe
```

**Comment by bpoplauschi** [comment]
Thanks @smindia1988 for sharing these settings

**Comment by jokerYellow** [comment]
great work.Thanks @smindia1988

**Comment by bpoplauschi** [comment]
Ok. I added https://github.com/SDWebImage/SDWebImage/wiki/Common-Problems#configuration-for-large-images-avoid-high-memory-usage

**Comment by ACode-Farmer** [comment]
@smindia1988 
I am using version 5.0.1 that don't has shouldDecompressImages and shouldDecompressImages.
So how to avoid crash on heavy images?

**Comment by dreampiggy** [comment]
@ACode-Farmer Use [.avoidDecodeImage](https://sdwebimage.github.io/Enums/SDWebImageOptions.html#/c:@E@SDWebImageOptions@SDWebImageAvoidDecodeImage). You can also use the OptionsProcesser for global control instead of invidual image request control. See #2736 

I'll update the Wiki soon about this. Seems out of date.

**Comment by ACode-Farmer** [comment]
Thanks for your  reply. @dreampiggy 
When I load multiple pictures, the memory is still too high.Use the same website for all pictures.(imageUrl = http://qiniu.ve-link.com//img/20191007/16d4d2e4e823e7449014aff5074096508372c2e9.jpg?imageView2/3/q/100).
SDWebImage  @config:
SDImageCache.sharedImageCache.config.maxDiskAge = 3600 * 24 * 7; // 1 week
    SDImageCache.sharedImageCache.config.maxMemoryCost = 1024 * 1024 * 20; //20 images
    SDImageCache.sharedImageCache.config.shouldUseWeakMemoryCache = false; //Default True => Store images in RAM cache for Fast performance
    SDImageCache.sharedImageCache.config.diskCacheReadingOptions = NSDataReadingMappedIfSafe;
SDWebImage Use:
[imageView sd_setImageWithURL:[NSURL URLWithString:obj] placeholderImage:nil options:SDImageCacheAvoidDecodeImage];
I tested three options setting: 
1、SDWebImageScaleDownLargeImages
2、SDWebImageAvoidDecodeImage
3、SDWebImageScaleDownLargeImages | SDWebImageAvoidDecodeImage
Memory is still not reduced.

**Comment by dreampiggy** [comment]
RAM usage for bitmap image usage, the fomula is really simple:

Pixel Count * 4.

For you demo image, it's a `3840px * 5120px`, is 75MB.

So, you'd better choose another idea for this, such as preprocess the image with lower resolution, or using something on the Client like [Wiki - Image Transformer](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#image-transformer-50). To scale down the resolution for images.

You can also try to use the [CATiledLayer](https://developer.apple.com/documentation/quartzcore/catiledlayer) to display super large images. Which means, you don't load the image entirelly into memory, you load them by tiles. Then, track the user's scroll and load another tile, etc. This is just the technolegy used by Map apps (For example, a full map bitmap image is 10000000*10000000 pixels, this can not been hold in RAM at all).

**Comment by dreampiggy** [comment]
@ACode-Farmer Update the documentation, see [Wiki - Optimization for large images](https://github.com/SDWebImage/SDWebImage/wiki/Common-Problems#optimization-for-large-images)

**Comment by ACode-Farmer** [comment]
Thanks @dreampiggy 
Which version are these settings in?

**Comment by dreampiggy** [comment]
The documentation is updated to latest SDK name. The options processor is available in 5.1.0, the other is available in 5.0.0. See wiki: https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#options-processor-510
